### PR TITLE
macOS: Add additional flags to OSLIBS

### LINKED ISSUE #49976: Alternative to copying part of a recent Xcode `libSystem.tbd` to `src/mach_dyld_atfork.tbd`
I find copying part of a recent Xcode `libSystem.tbd` file into the Julia source code rather odd.

It also prevents MacPorts from building current Julia versions (≥ 1.8.0) on 'OS X'/macOS 10.10 - 10.14 (Yosemite - Mojave).  

---

@quentinmit was able to get Julia to build on macOS 10.14 (Mojave)

> [...] with a simple Makefile patch; I replaced
> 
> ```
> OSLIBS += $(SRCDIR)/mach_dyld_atfork.tbd
> ```
> 
> with
> 
> ```
> CLANG_LDFLAGS += -Wl,-U,__dyld_atfork_prepare -Wl,-U,__dyld_atfork_parent
> ```
> 
> and added `$(CLANG_LDFLAGS)` to the rules building libjulia-internal and libjulia-internal-debug. The latter is probably safe on every platform but the former needs a guard for older OS versions, of course.
> 
> It's possible that the additional `-U` flags are unnecessary, but I didn't test that (the previous line sets a different flag that theoretically should allow unknown symbols as well).

– [#65550 (julia: link fails for 10.14 and earlier: ld: unsupported tapi file type '!tapi-tbd' in YAML file) – MacPorts](https://trac.macports.org/ticket/65550#comment:16)

---

Cross references:

* https://github.com/JuliaLang/julia/pull/43701
  * Fixed https://github.com/JuliaLang/julia/issues/43578
* https://github.com/JuliaLang/julia/pull/49740
  * Fixed https://github.com/JuliaLang/julia/issues/49733

### PR DESCRIPTION
**Pull request by benz0li**
- Remove file src/mach_dyld_atfork.tbd
- Fixes #49976

### CODE REVIEW

**Comment by benz0li** [comment]
Cross reference: https://github.com/macports/macports-ports/pull/18905

Same patch. But only applied if darwin version < 19 (macOS version < 10.15 (Catalina)).

**Comment by benz0li** [comment]
@quentinmit The MacPorts build for macOS 10.14 (Mojave) now succeeds.  
ℹ️ It does not build for older versions ('OS X'/macOS 10.10 - 10.14 (Yosemite - High Sierra)) on MacPorts, though.

File: src/Makefile [hunk 1]
```diff
@@ -149,16 +149,15 @@ endif
 
 CLANG_LDFLAGS := $(LLVM_LDFLAGS)
 ifeq ($(OS), Darwin)
-CLANG_LDFLAGS += -Wl,-undefined,dynamic_lookup
-OSLIBS += $(SRCDIR)/mach_dyld_atfork.tbd
+CLANG_LDFLAGS += -Wl,-undefined,dynamic_lookup -Wl,-U,__dyld_atfork_parent -Wl,-U,__dyld_atfork_prepare -Wl,-U,__dyld_dlopen_atfork_parent -Wl,-U,__dyld_dlopen_atfork_prepare -framework CoreFoundation
```
Comment (vtjnash) [inline src/Makefile:152]: I don't see how this is clang related, or why you are adding the clang flags to all compilers.

```suggestion
OSLIBS += -Wl,-undefined,dynamic_lookup -Wl,-U,__dyld_atfork_parent -Wl,-U,__dyld_atfork_prepare -Wl,-U,__dyld_dlopen_atfork_parent -Wl,-U,__dyld_dlopen_atfork_prepare -framework CoreFoundation
```

File: src/Makefile [hunk 2]
```diff
@@ -149,16 +149,15 @@ endif
...
 LIBJULIA_PATH_REL := @rpath/libjulia
 else
 LIBJULIA_PATH_REL := libjulia
 endif
 
 COMMON_LIBPATHS := -L$(build_libdir) -L$(build_shlibdir)
-RT_LIBS := $(WHOLE_ARCHIVE) $(LIBUV) $(WHOLE_ARCHIVE) $(LIBUTF8PROC) $(NO_WHOLE_ARCHIVE) $(LIBUNWIND) $(RT_LLVMLINK) $(OSLIBS) $(LIBTRACYCLIENT) $(LIBITTAPI)
-CG_LIBS := $(LIBUNWIND) $(CG_LLVMLINK) $(OSLIBS) $(LIBTRACYCLIENT) $(LIBITTAPI)
+RT_LIBS := $(WHOLE_ARCHIVE) $(LIBUV) $(WHOLE_ARCHIVE) $(LIBUTF8PROC) $(NO_WHOLE_ARCHIVE) $(LIBUNWIND) $(RT_LLVMLINK) $(LIBTRACYCLIENT) $(LIBITTAPI)
+CG_LIBS := $(LIBUNWIND) $(CG_LLVMLINK) $(LIBTRACYCLIENT) $(LIBITTAPI)
```
Comment (vtjnash) [inline src/Makefile:160]: You cannot just remove OSLIBS from the link line. Please remove this line from the diff.

File: src/Makefile (hunk 1 above)
Comment (benz0li) [reply src/Makefile]: Agreed. I moved the flags to `OSLIBS`.

File: src/Makefile (hunk 2 above)
Comment (benz0li) [reply src/Makefile]: Done.

**Comment by benz0li** [comment]
@vtjnash I am currently running the MacPorts build for macOS 10.14 (Mojave) locally.

If it succeeds, this can be merged.

File: src/Makefile
Comment (benz0li) [inline src/Makefile:1]: With `OSLIBS` present again, this is sufficient.

ℹ️ The MacPorts build for macOS 10.13 (High Sierra) succeeded.

Comment (benz0li) [reply src/Makefile:1]: Now, all checks pass.

**Comment by vtjnash** [comment]
It looks like it has the wrong version of LLVM, from the failure in that build log. Not sure how that could have happened though. Yes, linux is normally built with gcc (though can be built with clang).

**Pull request merged**